File: oneview/robotFwEngine/lib/sshConnector.py
# Author: Madhu Chakravarthy
# Date: 12-04-2017

import os
import json
import paramiko
import logging

#Initialize logger
logger = logging.getLogger('appLogger')

class SSHConnector(object):

    def __init__(self, host):

        logger.info("Initializing SSHConnector for host" + host)
        self.loginDetails = self.readHostInfo(host)
        self.connection = None

    def readHostInfo(self, host):

        logger.debug("Reading " + host + " details")
        filePath = os.path.join(os.getcwd(),'..', 'config', host + '.json')
        with open(filePath) as jsonFile:
            data = json.load(jsonFile)
        return data

    def connectToHost(self):

        logger.debug("Connecting to " + self.loginDetails['host'])
        self.connection = paramiko.SSHClient()
        self.connection.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        self.connection.connect(self.loginDetails['host'],
                                username=self.loginDetails['username'],
                                password=self.loginDetails['password'])

    def executeCommandInHost(self, command):

        if command:
            logger.debug("Executing command in " + self.loginDetails['host'])
            stdin, stdout, stderr = self.connection.exec_command(command)
            stdin.close()
            for line in stdout.read().splitlines():
                logger.debug('%s: %s' % (self.loginDetails['host'], line))
        else:
            logger.debug("Empty command given")

    def copyReportsFromHost(self, source, destination):

        if self.connection:
            logger.debug("Copying file to " + destination)
            sftp = self.connection.open_sftp()
            sftp.get(source, destination)

    def closeConnectionToHost(self):

        if self.connection:
            logger.debug("Closing connection to host " + 
                          self.loginDetails['host'])
            self.connection.close()

if __name__ == "__main__":

    import logging.config

    logging.config.fileConfig(os.path.join(os.getcwd(),'..',
                                          'config', 'logging.conf'))
    logger = logging.getLogger('appLogger')
    logger.info("Started testing SSHConnector")
    
    #sshConn = SSHConnector('10.0.0.3')
    sshConn = SSHConnector('192.168.2.9')
    sshConn.connectToHost()
    sshConn.executeCommandInHost('pybot /home/madhu/robot/demo.robot')
    sshConn.copyReportsFromHost('/home/madhu/robot/output.xml', '../reports/output.xml')
    sshConn.copyReportsFromHost('/home/madhu/robot/log.html', '../reports/log.html')
    sshConn.closeConnectionToHost()

File: ui/api/common/sshConnector.js
/*
Author: Madhu Chakravarthy
*/

// Initialize variables

var client = require('ssh2').Client;
var logger = require('./log').logger

var executeCommand = function(cmd, hostname, user, passwd) {

    logger.info("Executing command through ssh")
    var conn = client()
    var result = 'None'
    conn.on('ready', function() {
        logger.debug("Executing command: " + cmd)
        conn.exec(cmd, function(err, stream) {
            if (err) throw err;
            stream.on('close', function(code, signal) {
                logger.debug('Stream :: close :: code: ' + code + ', signal: ' + signal);
                conn.end();
            }).on('data', function(data) {
                logger.debug('STDOUT: ' + data);
                result = data
            }).stderr.on('data', function(data) {
                logger.error('STDERR: ' + data);
                result = data
            });
        });
    logger.info("Finished executing command through ssh")
    }).connect({
        host: hostname,
        port:22,
        username: user, 
        password: passwd 
    });
    return result
}
module.exports.executeCommand = executeCommand 
